### DISCUSSION TITLE
Feature request: Join validation

### DISCUSSION THREAD

**Discussion by bayre** [duckdb/duckdb#10615] [Ideas]
File this one under [friendlier SQL](https://duckdb.org/2023/08/23/even-friendlier-sql.html)! 

[Pandas](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html), [Dplyr](https://dplyr.tidyverse.org/reference/mutate-joins.html#value:~:text=and%20y%20independently.-,relationship,-Handling%20of%20the) and [Polars](https://docs.pola.rs/py-polars/html/reference/dataframe/api/polars.DataFrame.join.html) all have a `validate` option, or a variation thereon, that allows the user to check whether a join performs as expected (one-to-many, one-to-one etc). From the limited research I've done it looks as though this isn't a feature implemented in other SQL engines, but I think it would be a great fit for DuckDB. There are some interesting suggestions for possible syntax in this [StackOverflow thread](https://stackoverflow.com/questions/75398473/is-there-a-validate-the-cardinality-of-sql-joins).

**Comment by polvoazul** [comment]
Suggestion: Add a 'UNIQUE' join modifier


Say you are joining two tables A and B. 

Now, if you do:

`FROM A LEFT JOIN B USING (name)`

Lets focus on what happens in these cases where the relationship is:
MANY-ONE (every name on A appears 0 or once in B)
vs
MANY-MANY (every name on A appears 0 or more times in B)

In the MANY-MANY cases we can end up with more lines in the resulting set than in A.
I would say that there are lots of times in an analytic context where this is undesirable. You would rather get an error than duplicated lines. Especially if those lines are further aggregated down the line and the error is thus hidden from view.

My proposal is to add a optional modifier to JOIN

`FROM A UNIQUE LEFT JOIN B USING (name)`

This will GUARANTEE that the resulting set will be the same size as A or else error out.

Of course this could be used with INNER JOIN to further guarantee a join where each and every line has a single match. So output has the same size and no nulls.

**Comment by alippai** [comment]
This is not just validation. It adds better hash join implementation (known size, right size is 1), cardinality estimation.
